src/coreclr/jit/fgprofilesynthesis.cpp

```diff
@@ -0,0 +1,917 @@
+    bool const isNextThrow = (next->bbJumpKind == BBJ_THROW);
+
+    if (isJumpThrow != isNextThrow)
+    {
+        if (isJumpThrow)
+        {
+            jumpEdge->setLikelihood(0.0);
+            nextEdge->setLikelihood(1.0);
+        }
+        else
+        {
+            jumpEdge->setLikelihood(1.0);
+            nextEdge->setLikelihood(0.0);
+        }
+
+        return;
+    }
+
+    // LOOP BACK EDGE heuristic
+    //
+    bool const isJumpEdgeBackEdge = IsLoopBackEdge(jumpEdge);
+    bool const isNextEdgeBackEdge = IsLoopBackEdge(nextEdge);
+
+    if (isJumpEdgeBackEdge != isNextEdgeBackEdge)
+    {
+        if (isJumpEdgeBackEdge)
+        {
+            JITDUMP(FMT_BB "->" FMT_BB " is loop back edge\n", block->bbNum, jump->bbNum);
+            jumpEdge->setLikelihood(0.9);
+            nextEdge->setLikelihood(0.1);
+        }
+        else
+        {
+            JITDUMP(FMT_BB "->" FMT_BB " is loop back edge\n", block->bbNum, next->bbNum);
+            jumpEdge->setLikelihood(0.1);
+            nextEdge->setLikelihood(0.9);
+        }
+
+        return;
+    }
+
+    // LOOP EXIT EDGE heuristic
+    //
+    // Consider: adjust probability if loop has multiple exit edges, so that
+    // overall exit probability is around 0.1.
+    //
+    bool const isJumpEdgeExitEdge = IsLoopExitEdge(jumpEdge);
+    bool const isNextEdgeExitEdge = IsLoopExitEdge(nextEdge);
+
+    if (isJumpEdgeExitEdge != isNextEdgeExitEdge)
+    {
+        if (isJumpEdgeExitEdge)
+        {
+            JITDUMP(FMT_BB "->" FMT_BB " is loop exit edge\n", block->bbNum, jump->bbNum);
+            jumpEdge->setLikelihood(0.1);
+            nextEdge->setLikelihood(0.9);
+        }
+        else
+        {
+            JITDUMP(FMT_BB "->" FMT_BB " is loop exit edge\n", block->bbNum, next->bbNum);
+            jumpEdge->setLikelihood(0.9);
+            nextEdge->setLikelihood(0.1);
+        }
+
+        return;
+    }
+
+    // RETURN heuristic
+    //
+    bool const isJumpReturn = (jump->bbJumpKind == BBJ_RETURN);
+    bool const isNextReturn = (next->bbJumpKind == BBJ_RETURN);
+
+    if (isJumpReturn != isNextReturn)
+    {
+        if (isJumpReturn)
+        {
+            jumpEdge->setLikelihood(0.2);
+            nextEdge->setLikelihood(0.8);
+        }
+        else
+        {
+            jumpEdge->setLikelihood(0.8);
+            nextEdge->setLikelihood(0.2);
+        }
+
+        return;
+    }
+
+    // IL OFFSET heuristic
+    //
+    // Give slight preference to bbNext
+    //
+    jumpEdge->setLikelihood(0.48);
+    nextEdge->setLikelihood(0.52);
+}
+
+//------------------------------------------------------------------------
+// AssignLikelihoodCond: update edge likelihood for a block that
+//   ends in a switch
+//
+// Arguments;
+//   block -- block in question (BBJ_SWITCH)
+//
+void ProfileSynthesis::AssignLikelihoodSwitch(BasicBlock* block)
+{
+    // Assume each switch case is equally probable
+    //
+    const unsigned n = block->NumSucc();
```

Comment:
```suggestion
    const unsigned n = block->NumSucc(m_comp);
```
?

Reply:
Ok, makes sense. Might want to add a comment to ensure that is perfectly clear, since we normally pair `NumSucc()/GetSucc()/Succs()` and `NumSucc(Compiler*)/GetSucc(Compiler*,x)/Succs(Compiler*)` and seeing them "unbalanced" is weird.